orc8r/cloud/go/services/analytics/collector_servicer.go

```diff
@@ -18,33 +18,132 @@ import (
+	calculations []calculations.Calculation,
+	userStateManager calculations.UserStateManager) protos.AnalyticsCollectorServer {
+	return &CollectorServicer{
+		promAPIClient:    promAPIClient,
+		calculations:     calculations,
+		analyticsConfig:  analyticsConfig,
+		userStateManager: userStateManager,
+	}
 }
 
-//Collect does the operation of running through calculations and returning results
-func (svc *CollectorService) Collect(context.Context, *protos.CollectRequest) (*protos.CollectResponse, error) {
+// Collect does the operation of running through calculations and returning results
+func (c *CollectorServicer) Collect(context.Context, *protos.CollectRequest) (*protos.CollectResponse, error) {
+	if c.userStateManager != nil {
+		c.userStateManager.Update()
+	}
+
 	response := &protos.CollectResponse{}
-	for _, calc := range svc.calculations {
-		results, err := calc.Calculate(svc.promAPIClient)
+	for _, calc := range c.calculations {
+		results, err := calc.Calculate(c.promAPIClient)
 		if err != nil {
-			glog.Errorf("Error calculating metric: %s", err)
+			glog.Errorf("Error %v calculating metric for %v", err, calc.GetCalculationParams())
 			continue
 		}
-		glog.V(10).Infof("results are %v", results)
-		response.Results = append(response.Results, results...)
+
+		c.registerResults(calc.GetCalculationParams(), results)
+		filteredResults := c.filterResults(results)
+		response.Results = append(response.Results, filteredResults...)
 	}
-	glog.V(10).Infof("response is %v", response)
 	return response, nil
 }
+
+// FilterResults filters the results to ensure that the user aggregate metrics
+// get filtered out in case the number of users in that context falls below
+// minimum threshold
+func (c *CollectorServicer) filterResults(results []*protos.CalculationResult) []*protos.CalculationResult {
+	userStateManager := c.userStateManager
+	// Nothing to filter
+	if userStateManager == nil {
+		return results
+	}
+
+	filteredResults := []*protos.CalculationResult{}
+	analyticsConfig := c.analyticsConfig
+	for _, result := range results {
+		// Enforce the minimum user threshold constraint at network level, gateway level and deployment level
+		metricConfig, ok := analyticsConfig.Metrics[result.GetMetricName()]
+		if !ok {
+			glog.Errorf("Metric Configuration not found for %s", result.GetMetricName())
+			continue
+		}
+		if !metricConfig.Export {
+			glog.V(1).Infof("Metric %s export configuration disabled", result.GetMetricName())
+			continue
+		}
+		if !metricConfig.EnforceMinUserThreshold {
+			filteredResults = append(filteredResults, result)
+			continue
+		}
+		networkID, networkLabelPresent := result.Labels[metrics.NetworkLabelName]
+		gatewayID, gatewayLabelPresent := result.Labels[metrics.GatewayLabelName]
+
+		if networkLabelPresent {
+			if gatewayLabelPresent {
+				numUsersInGateway := userStateManager.GetTotalUsersInGateway(networkID, gatewayID)
+				if numUsersInGateway > analyticsConfig.MinUserThreshold {
```

Comment:
Oh one last thing -- see my previous comment on this. I still think this nesting can be improved a bit. There's almost definitely a function you can abstract out to handle e.g. lines 102-107